homeassistant/components/sensor/rfxtrx.py

```diff
@@ -46,13 +47,16 @@ def setup_platform(hass, config, add_devices_callback, discovery_info=None):
                 if data_type in event.values:
                     data_types = [data_type]
                     break
+        add_devices_callback(sensors)
         for _data_type in data_types:
-            new_sensor = RfxtrxSensor(None, entity_info[ATTR_NAME],
-                                      _data_type, entity_info[ATTR_FIREEVENT])
-            sensors.append(new_sensor)
+            entity_id = generate_entity_id(ENTITY_ID_FORMAT,
+                                           entity_info[ATTR_NAME],
+                                           hass=hass)
+            new_sensor = RfxtrxSensor(None, entity_info[ATTR_NAME], _data_type,
+                                      entity_info[ATTR_FIREEVENT], entity_id)
             sub_sensors[_data_type] = new_sensor
+            add_devices_callback([new_sensor])
```

Comment:
Add devices takes in a list, that means that the order will always remain the same as it was passed in. It is a big performance hit to move to 1 list per device being passed in. 

So if the order is not deterministic, it might be related to the order the devices get discovered in by rfxtrx: the values inside `data_types`.